feature(flyTo): Fly to the extent of mvt entry type layers with featureSet/Lookup

AlexanderGeere opened the pull request:
## Description
This PR makes it possible to use the `flyTo` to fly to the greater extent of a location which uses an entry type layer, and featureSet or featureLookup. 

Previously `flyTo` would zoom in only on the location itself:
![Screenshot from 2025-06-27 11-47-48](https://github.com/user-attachments/assets/0cb08635-703d-4f22-af0c-c9fde1811696)

On the PR it should now flyTo the outer extent:
![Screenshot from 2025-06-27 11-49-06](https://github.com/user-attachments/assets/0ab05526-b3a7-4411-a958-628f901d94ff)

This is done by adding a `getExtent` function to the pin, geometry and layer entry. These functions are added to an Extents object on the location, when the entry is shown. The new `createExtent` function will loop over this object and create an extent which contains all the extens present in location.Extents.

`flyTo` calls `createExtent` with the location.Extents object. 

## GitHub Issue
#1469

## Type of Change
- ✅ New feature (non-breaking change which adds functionality)
- ✅ Documentation

## How have you tested this?
Tested locally using `bugs_testing/infoj/layer_entry.json` on the points layer.
[Test Location](http://localhost:3000/latest/?layers=OSM,uk_retail_places,points&locations=points!4&z=10.03&lat=51.54777&lng=-0.05503)


## Testing Checklist
- ✅ Existing Tests still pass
- ✅ Ran locally on my machine
- ✅ Updated existing tests
- ✅ Added new tests

## Code Quality Checklist
- ✅ My code follows the guidelines of XYZ
- ✅ My code has been commented
- ✅ Documentation has been updated
- ✅ New and existing unit tests pass locally with my changes
- ✅ Main has been merged into this PR

dbauszus-glx reviewed (dismissed):
I don't understand the change to the feature_extent query template. Why is this required? Doesn't this mess zoomToExtent method in the layer decorate module?

Having a getFeatureExtent method with duplicate code to the zoomToExtent method in the location layer entry should be avoided.

The location layer entry method uses the layer decorator. The getExtent method should be added to the layer decorator. The zoomToExtent method can then call the internal getExtent method.

I'd probably split the location flyTo method. Have a location getExtent method called from the flyTo method.

What is the `values_` property of the entry.L object in the location.layers array. This looks like an internal OL property which should not be accessed directly.

I'd probably rethink the use of the location.Layers array. The idea behind this array is to provide a way to access the extent of multiple layer and to ensure that the layer is removed when the location is removed. This predates the removeCallbacks array which can be used to remove the Openlayers layer from the mapview.

It is probably easier to rethink the getExtent for geometry / pin entries than having a complex condition in the flyTo / getExtent methods for a location.

AlexanderGeere commented:
@dbauszus-glx I have added a  `getExtent` to the pin, geometry and layer entries. I have also added an Extents property to the location.

Extent functions are added to the Extents parameter when an entry is shown and removed when it is hidden. 

When clicking `flyTo`, `createExtent` is called with the location.Extents. Create extent returns an extent that includes all those from location.Extents. 

On the pin entry `getExtent` function, a buffer can be applied to the pin by either supplying `entry.buffer: true` or `entry.buffer: 500` for instance.

dbauszus-glx reviewed (changes requested):
The zoom to location breaks on a vector layer.

```
          {
            "type": "layer",
            "layer": "points"
          }
```

The points layer is a WKT vector layer.

![image](https://github.com/user-attachments/assets/e904f547-435a-4c93-8150-e4a23fbb3099)

layer.mjs:264 Uncaught (in promise) TypeError: e.getExtent is not a function
    at location.Extents.<computed> [as points-327] (layer.mjs:264:63)
    at Object._e [as createExtent] (createExtent.mjs:17:61)
    at async Object.un [as flyTo] (decorate.mjs:292:23)

AlexanderGeere commented:
@dbauszus-glx  the getExtent function is now always asssigned to the layer entry and teh check on featureSet or lookup is done in the function

dbauszus-glx commented:
The getExtent method should be assigned to the layer in the decorator. The layer will be this in the method.

The getFeatureExtent query should return values only.

I have also updated the conditions inside the getExtent method in regards to the check for the featureLookup, featureSet, or else.

https://github.com/GEOLYTIX/xyz/pull/2235/commits/e05e2c4186784a87f022b48e5ab1f6dc28d13cda

dbauszus-glx commented:
The location.extents should not be processed in the layer.show method assigned in the layer decorator module. The layer entry module assigns it's own showLayer method. https://github.com/GEOLYTIX/xyz/pull/2235/commits/38c8c1f24ef29834072fe92b0b554d8469c2b113

dbauszus-glx commented:
There was quite a bit of duplicate code and some broken SQL.

The layer_extent SQL could not work due to a missing SELECT in the nested query and a surplus semi-colon.
![image](https://github.com/user-attachments/assets/b90cb45e-c944-4370-b3bc-1a58e3f2823f)

Having a feature_extent and a layer_extent query which are identical apart from the body condition is too much duplicate code.

This can be solved by creating the condition string inside the template in regards to the body with an id array being available.

The query method still needs some documentation.

```js
export default (_) => {

  let condition = 'true'

  if (Array.isArray(_.body?.ids)) {
    const ids = _.body.ids.filter((id) => /^[A-Za-z0-9,"'._-\s]*$/.test(id));

    condition = `\${qID} IN ('${ids.join("','")}')`
  }

  return `
    SELECT
      ARRAY[
        st_xmin(box.box2d),
        st_ymin(box.box2d),
        st_xmax(box.box2d),
        st_ymax(box.box2d)]
    FROM(
      SELECT Box2D(
        ST_Transform(
          ST_SetSRID(
            ST_Extent(\${geom}),
            \${proj}),
          \${srid}))
      FROM \${table}
      WHERE ${condition} \${filter}) box`
};
```

The query has been taken out from the zoomToExtent method. The getExtent method should now return the Extent for any type of layer with one query. The body is provided as undefined if there are no ids.

```js
async function getExtent() {
  const layer = this;

  let ids = [];
  if (Array.isArray(layer.featureLookup)) {
    // Get array of ids from lookup features.
    ids = layer.featureLookup.map(
      (feature) => feature[layer.featureLookupId || 'id'],
    );
  } else if (Array.isArray(layer.featureSet)) {
    ids = [...layer.featureSet];
  } else if (layer.L && layer.format !== 'mvt') {
    // return extent from Openlayers source.
    return layer.L.getSource().getExtent();
  }

  const body = ids.length? JSON.stringify({
      ids: ids,
    }) : undefined;

  const response = await mapp.utils.xhr({
    url:
      `${layer.mapview.host}/api/query/layer_extent?` +
      mapp.utils.paramString({
        filter: layer.filter?.current,
        geom: layer.geomCurrent(),
        layer: layer.source_key || layer.key,
        locale: layer.mapview.locale.key,
        proj: layer.srid,
        srid: layer.mapview.srid,
        table:
          layer.table ||
          Object.values(layer.tables)[0] ||
          Object.values(layer.tables)[1],
      }),
    body,
  });

  return response;
}
```

The feature_extent test is now breaking and needs to be updated. @AlexanderGeere

dbauszus-glx commented:
@AlexanderGeere Please check my last two commits.

https://github.com/GEOLYTIX/xyz/pull/2235/commits/05e42ffb81475efa2017938c22a68a6fe56f922a

Instead of an utility method to calculate the extents from an object of extents it is easier to have a location.getExtent method which calculates it's extent.

The getExtent method of pin, geometry, and layer entry should be assigned as a method where `this` is the entry. The entry does not have to be provided to this methods.

It is much easier to assign the getExtent method once rather than to assign and reassign in the show hide methods.

https://github.com/GEOLYTIX/xyz/pull/2235/commits/858456a66f69b7c203b085685888a076dd66307d

The location.Extents object is no longer required. The layer [entry] have each show, hide, getExtent methods and can be pushed into the Layers array of the location. You can now easily iterate over that array to call the remove method of each entry, or the hide method of each entry.

simon-leech reviewed (dismissed):
This is looking really good! 
A few little issues:
1. Zooming to the extent of just a `pin` entry, isn't working, it doesn't zoom in. 
2. Calling `flyTo()` in a custom view on a location is not doing anything. I think this may be to do with the flyTo requiring a `maxZoom` but we never previously had to define this, it had a default.

AlexanderGeere commented:
Added a entry.display ??= true on the pin, without this the flyTo does not fly to the pin if it is the only geometry on the layer.

cityremade commented:
I noticed a behaviour on styled catchments when at the end of zoom tiles are not recreated. How does it work when map zoom is finished and mvt tiles need to be refreshed for the final zoom level?
